### PR TITLE
[http] Initial contribution

### PR DESCRIPTION
**Pull request by J-N-K**
This adds a new HTTP binding as a replacement for the old http1 binding.

Replaces #4342 

Signed-off-by: Jan N. Klug <jan.n.klug@rub.de>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
@J-N-K Thanks for picking that (very important) binding up!
I am a bit puzzled though that your PR has a new skeleton as a basis - I hope you will base the work here on the code that has been implemented in #4342 and into which some considerable effort already went?

**Comment by J-N-K** [comment]
I did chose a completely different implementation approach but did take into account the comments regarding the user-facing thing/channel structure.

**Comment by bodiroga** [comment]
@J-N-K, what happened here? Why did you close the PR? :disappointed:

**Comment by J-N-K** [comment]
@bodiroga I didn't want to "steal" the PR from someone else. 

I did chose not to continue the work from #4342 and restart because I did look through the code and didn't get the idea behind some code design decisions. At some points (like auth) the other PR decided to do thing with own methods/classes where similar fucntionality can be gained by using methods provided by the Jetty client itself. The PR is also based on the old build system and would need migration.

Since it's probably me who needs to maintain and fix issues after merge, I thought it would be better to take the user-facing design (i.e. thing/channel) that was discussed in that PR from scratch. It was not meant as a hostile takeover or a fuck-off to @kelnos.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpThingHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpThingHandler.java:183]: I hate code duplication, but even I don't know if what I'm suggesting is worth it...

```suggestion
                itemValueConverter = createConverter(ColorItemConverter::new, channelUID, channelConfig);
```

```
public ItemValueConverter createConverter(ItemConverterFactory factory, ChannelUID channelUID,
HttpChannelConfig channelConfig){
        String stateUrl = concatenateUrlParts(config.baseURL, channelConfig.stateExtension);
        String commandUrl = channelConfig.commandExtension == null ? stateUrl
                : concatenateUrlParts(config.baseURL, channelConfig.commandExtension);

     return factory.create(state -> updateState(channelUID, state),
                        command -> sendCommand(commandUrl, command),
                        valueTransformationProvider.getValueTransformation(channelConfig.stateTransformation),
                        valueTransformationProvider.getValueTransformation(channelConfig.commandTransformation),
                        channelConfig);
}


@FunctionalInterface
private interface ItemConverterFactory{
    ItemValueConverter create(Consumer<State> updateState, @Nullable Consumer<String> sendCommand,
            ValueTransformation stateTransformations, ValueTransformation commandTransformations,
            HttpChannelConfig channelConfig);
}
```

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/ColorItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/ColorItemConverter.java:88]:
```suggestion
                    new PercentType(((HSBType) state).getBrightness().toBigDecimal().subtract(channelConfig.step)));
```

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/ColorItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/ColorItemConverter.java:85]: need to make sure that max values are not exceeded

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/DimmerItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/DimmerItemConverter.java:76]: You should make `BigDecimal.valueOf(100)` a constant

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/RollershutterItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/RollershutterItemConverter.java:57]:
```suggestion
            if (command.equals(PercentType.HUNDRED) && !channelConfig.downValue.isEmpty()) {
```

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/RollershutterItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/RollershutterItemConverter.java:59]:
```suggestion
            } else if (command.equals(PercentType.ZERO) && !channelConfig.upValue.isEmpty()) {
```

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/RollershutterItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/RollershutterItemConverter.java:101]:
```suggestion
        try {
            BigDecimal value = new BigDecimal(string);
            if (value.compareTo(PercentType.HUNDRED.toBigDecimal()) > 0) {
                return PercentType.HUNDRED;
            }
            if (value.compareTo(PercentType.ZERO.toBigDecimal()) < 0) {
                return PercentType.ZERO;
            }
            return new PercentType(value);
        } catch (NumberFormatException e) {
        }
        
        return UnDefType.UNDEF;
```

Comment (J-N-K) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/RollershutterItemConverter.java:101]: I don't think that is equivalent, the internal state is not updated.

Comment (cpmeister) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/RollershutterItemConverter.java:101]: Oof, you are right, I was getting a bit too ahead of myself. Leave as is then.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpThingHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpThingHandler.java:281]: death to duplicate code.
```suggestion
    private <T extends State> ItemValueConverter createGenericConverter(String commandUrl,
            ChannelUID channelUID, HttpChannelConfig channelConfig, Class<T extends State> clazz) {
        AbstractTransformingItemConverter.Factory factory = (state, command, value, stateTrans, commandTrans, config) -> new GenericItemConverter<T>(clazz, state, command, value, stateTrans, commandTrans, config);
        return createTransformingConverter(factory, commandUrl, channelUID, channelConfig);
```

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/GenericItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/GenericItemConverter.java:66]: With my change above, this interface is obsolete and can be removed.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/config/HttpChannelConfig.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/config/HttpChannelConfig.java]:
```
private void addToMaps(@Nullable String value, State state){
    if (value != null) {
        commandStringMap.put(state, value);
        stringStateMap.put(value, state);
    }
}
```
```suggestion
        addToMaps(this.onValue, OnOffType.ON);
        addToMaps(this.offValue, OnOffType.OFF);
        addToMaps(this.openValue, OpenClosedType.OPEN);
        addToMaps(this.closedValue, OpenClosedType.CLOSED);
        addToMaps(this.upValue, UpDownType.UP);
        addToMaps(this.downValue, UpDownType.DOWN);
```

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/ColorItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/ColorItemConverter.java:42]: The parent class already has this field, so you don't need to define it here.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/DimmerItemConverter.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/DimmerItemConverter.java:36]: same here

Comment (J-N-K) [reply bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/converter/DimmerItemConverter.java:36]: The reason I defined it again is that the compiler complains about assigning values to a final variable. Removed the `final` in the base class.

File: bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/HttpResponseListener.java:50]:
```suggestion
    private final String fallbackEncoding = StandardCharsets.UTF_8.name();

    public HttpResponseListener(CompletableFuture<@Nullable Content> future) {
        this(future, null);
    }

    public HttpResponseListener(CompletableFuture<@Nullable Content> future, @Nullable String fallbackEncoding) {
        this.future = future;
        this.fallbackEncoding = fallbackEncoding != null ? fallbackEncoding : StandardCharsets.UTF_8.name();
    }
```

File: bundles/org.openhab.binding.http/src/main/resources/ESH-INF/i18n/http_xx_XX.properties
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/resources/ESH-INF/i18n/http_xx_XX.properties:4]: This file should be removed

File: bundles/org.openhab.binding.http/src/main/resources/ESH-INF/thing/thing-types.xml
Comment (cpmeister) [inline bundles/org.openhab.binding.http/src/main/resources/ESH-INF/thing/thing-types.xml:69]: This list is hardly exhaustive, should you allow users to input it manually instead?

File: bundles/org.openhab.binding.http/README.md
Comment (cpmeister) [inline bundles/org.openhab.binding.http/README.md:118]: Looks like examples still need to be added in.

File: bundles/org.openhab.binding.http/README.md
Comment (cpmeister) [inline bundles/org.openhab.binding.http/README.md:19]: You should fill in the `optional` column for every parameter here, even if it would be implied by the presence of a default value.

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.http/README.md
Comment (J-N-K) [reply bundles/org.openhab.binding.http/README.md:118]: Good question, is this still needed in OH3? We won't merge before.

**Comment by kaikreuzer** [comment]
Could you please re-create this PR against the `main` branch?
See point (4) in https://github.com/openhab/openhab-addons/issues/8512. Thanks!

**Pull request closed without merging**
